src/dune_rules/melange_rules.ml

```diff
@@ -0,0 +1,197 @@
+open Import
+module CC = Compilation_context
+
+let js_includes ~emit_stanza_dir ~target ~requires_link ~scope =
+  let open Resolve.Memo.O in
+  Command.Args.memo
+    (Resolve.Memo.args
+       (let+ (libs : Lib.t list) = requires_link in
+        let project = Scope.project scope in
+        let deps_of_lib (lib : Lib.t) =
+          let lib = Lib.Local.of_lib_exn lib in
+          let info = Lib.Local.info lib in
+          let lib_dir = Lib_info.src_dir info in
+          let dst_dir =
+            Melange.lib_output_dir ~emit_stanza_dir ~lib_dir ~target
+          in
+          [ (let dir = Path.build dst_dir in
+             Glob.of_string_exn Loc.none ("*" ^ Melange.js_ext)
+             |> File_selector.of_glob ~dir |> Dep.file_selector)
+          ]
+          |> Dep.Set.of_list
+        in
+        let deps libs = Dep.Set.union_map libs ~f:deps_of_lib in
+        Command.Args.S
+          [ Lib_flags.L.include_flags ~project libs Melange
+          ; Hidden_deps (deps libs)
+          ]))
+
+let build_js ?loc ~pkg_name ~module_system ~dst_dir ~obj_dir ~sctx ~build_dir
+    ~lib_deps_js_includes m =
+  let cm_kind = Lib_mode.Cm_kind.Melange Cmj in
+  let mode = Lib_mode.of_cm_kind cm_kind in
+  let open Memo.O in
+  let* compiler =
+    let+ compiler =
+      (* TODO loc should come from the mode field in the dune file *)
+      Super_context.resolve_program sctx ~loc:None ~dir:build_dir
+        ~hint:"opam install melange" "melc"
+    in
+    Result.to_option compiler
+  in
+  (let open Option.O in
+  let+ compiler = compiler in
+  let src = Obj_dir.Module.cm_file_exn obj_dir m ~kind:cm_kind in
+  let in_dir = Path.Build.relative dst_dir in
+  let output =
+    let name =
+      Module_name.Unique.artifact_filename (Module.obj_name m)
+        ~ext:Melange.js_ext
+    in
+    in_dir name
+  in
+  let obj_dirs =
+    Obj_dir.all_obj_dirs obj_dir ~mode
```

Comment:
Shouldn't this just be the melange dir? Or are the cmi's needed too?

Reply:
Right, removed in https://github.com/ocaml/dune/pull/6349/commits/18d987238858e0a9950632fc453c03004937cc74.